### PR TITLE
gh-810 Add support for RBAC over REST-APIs

### PR DESCRIPTION
**Pull request by ghillert**
* Ensure the Ldap Tests work with roles

Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/810
Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/1110

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java [hunk 1]
```diff
@@ -123,4 +163,55 @@ protected void configure(HttpSecurity http) throws Exception {
...
+		return security;
+	}
+
+	/**
+	 * Holds configuration for the authorization aspects of security.
+	 *
+	 * @author Eric Bottard
+	 * @author Gunnar Hillert
+	 */
+	@ConfigurationProperties(prefix = "spring.dataflow.security.authorization")
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java]: Should this be `spring.cloud.dataflow.security.authorization` ?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java [hunk 2]
```diff
@@ -123,4 +163,55 @@ protected void configure(HttpSecurity http) throws Exception {
...
+					String.format(
+							"Unable to parse security rule [%s], expected format is 'HTTP_METHOD ANT_PATTERN => SECURITY_ATTRIBUTE(S)'",
+							rule));
+
+			HttpMethod method = HttpMethod.valueOf(matcher.group(1).trim());
+			String urlPattern = matcher.group(2).trim();
+			String attribute = matcher.group(3).trim();
+
+			logger.error(">>>>Authorization '{}' | '{}' | '{}'", method, attribute, urlPattern);
+			security = security.antMatchers(method, urlPattern).access(attribute);
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java]: User with `ROLE_ADMIN` privilege can't access the URL mapped under `ROLE_CREATE`. For instance, user with `ROLE_ADMIN` can't register apps as the registration POST URL is mapped for `ROLE_CREATE`.

Comment (ghillert) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java]: `ROLE_ADMIN` is strictly for the management endpoints. Maybe we should rename it to `ROLE_MANAGEMENT`?

Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java]: `ROLE_ADMIN` inherently sounds like user with all privileges. If we want to restrict access only to management endpoints, then explicitly specifying `ROLE_MANAGEMENT_ENDPOINTS` makes more sense than just `ROLE_MANAGEMENT` as the later also closely aligns with admin IMO.

Comment (ericbottard) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java]: +1 to what Ilaya says, although I would call it `ROLE_MANAGE` to better align with other role names.
ROLE_ADMIN should be the almighty god IMO

**Comment by ghillert** [comment]
@ilayaperumalg  Added another commit:

* Add reference doc for role-based access control
* Code Review Changes
  - Standardize on / use `spring.cloud.dataflow` prefix for ALL relevant security YAML properties

@markpollack @sabbyanandan How important is it to avoid breaking changes? Right now, due to the change of properties, there is a minor security-related breaking change. If needed I could rework the code to provide the:

- old properties with deprecation warnings
- **and** the new properties

File: spring-cloud-dataflow-docs/src/main/asciidoc/configuration.adoc [hunk 3]
```diff
@@ -400,6 +406,135 @@ tracker ticket] to track its implementation. You may also want to look at the
...
+part of the Spring Cloud Data Flow Core Module. Nonetheless, you can
+override those, if desired, e.g. in `application.yml`. The configuration takes
+the form of a YAML *list* (as some rules may have precedence over others) and so
+you'll need to copy/paste the whole list and tailor it to your needs (as there is no way to merge lists). Always refer to your version of `application.yml`, as the snippet reproduced
+below may be out-dated. The default rules are as such:
+
+[source,yaml]
+----
+spring:
+  ...
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-docs/src/main/asciidoc/configuration.adoc]: I guess we can mention `cloud` here explicitly without ...

**Comment by sabbyanandan** [comment]
> How important is it to avoid breaking changes? Right now, due to the change of properties, there is a minor security-related breaking change. If needed I could rework the code to provide the:

Given this is a new feature altogether and the fact that we are consolidating the naming conventions, I'd vote for the breaking change. Also, this feature is targeted for 1.2.x release line, so we could capture this in the release upgrade notes. 

@markpollack - what are your thoughts to: _"Standardize on / use `spring.cloud.dataflow` prefix"_?

File: spring-cloud-dataflow-docs/src/main/asciidoc/configuration.adoc [hunk 4]
```diff
@@ -400,6 +406,135 @@ tracker ticket] to track its implementation. You may also want to look at the
...
+* SECURITY_ATTRIBUTE is a SpEL expression (see http://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#el-access)
+* each of those separated by one or several blank characters (spaces, tabs, _etc._)
+
+Be mindful that the above is indeed a YAML list, not a map (thus the use of '-' dashes at the start of each line) that lives under the `spring.cloud.dataflow.security.authorization.rules` key.
+
+[TIP]
+====
+In case you are solely interested in authentication but not authorization,
+for instance every user shall have have access to all endpoints, then you can also
+set `spring.cloud.dataflow.security.authorization.enabled=true`.
```
Comment (ericbottard) [inline spring-cloud-dataflow-docs/src/main/asciidoc/configuration.adoc]: Should this be `=false` ?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java [hunk 5]
```diff
@@ -123,4 +163,55 @@ protected void configure(HttpSecurity http) throws Exception {
...
+			Assert.isTrue(matcher.matches(),
+					String.format(
+							"Unable to parse security rule [%s], expected format is 'HTTP_METHOD ANT_PATTERN => SECURITY_ATTRIBUTE(S)'",
+							rule));
+
+			HttpMethod method = HttpMethod.valueOf(matcher.group(1).trim());
+			String urlPattern = matcher.group(2).trim();
+			String attribute = matcher.group(3).trim();
+
+			logger.error("Authorization '{}' | '{}' | '{}'", method, attribute, urlPattern);
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/BasicAuthSecurityConfiguration.java]: why `error` here?

File: spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml [hunk 6]
```diff
@@ -1,3 +1,76 @@
...
+            # Task Definitions
+
+            - POST   /tasks/definitions              => hasRole('ROLE_CREATE')
+            - DELETE /tasks/definitions/*            => hasRole('ROLE_CREATE')
+            - GET    /tasks/definitions              => hasRole('ROLE_VIEW')
+            - GET    /tasks/definitions/*            => hasRole('ROLE_VIEW')
+
+            # Task Deployments
+
+            - POST   /tasks/deployments/*            => hasRole('ROLE_CREATE')
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: This does not exist anymore

File: spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml [hunk 7]
```diff
@@ -1,3 +1,76 @@
...
+
+            # Task Executions
+
+            - GET    /tasks/executions               => hasRole('ROLE_VIEW')
+            - GET    /tasks/executions/*             => hasRole('ROLE_VIEW')
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: BUT a POST exsits here now.
Also, a DELETE for task execution cleanup

File: spring-cloud-starter-dataflow-server-local/pom.xml [hunk 8]
```diff
@@ -33,6 +33,11 @@
 			<version>1.5.5</version>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
```
Comment (ericbottard) [inline spring-cloud-starter-dataflow-server-local/pom.xml]: What's the code change that actually needs this?

File: spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml (hunk 6 above)
Comment (ghillert) [inline spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: It is in the `TaskDeploymentController`

```
@RestController
@RequestMapping("/tasks/deployments")
@ExposesResourceFor(TaskDeploymentResource.class)
public class TaskDeploymentController {
```

File: spring-cloud-starter-dataflow-server-local/pom.xml (hunk 8 above)
Comment (ghillert) [inline spring-cloud-starter-dataflow-server-local/pom.xml]: In `LocalServerSecurityWithUsersFileTests` I use the Guava test dependency for creating Maps, e.g. to define test request parameter:

```
ImmutableMap.of("start", "2", "detailLevel", "2")
```

File: spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml (hunk 7 above)
Comment (ghillert) [inline spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: Added with latest commit

**Comment by ghillert** [comment]
Rebased and 

* Ensure `UnsatisfiedServletRequestParameterException` will trigger a `400` Http Status
* Handle POST and DELETE for `/tasks/executions`

File: spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml (hunk 6 above)
Comment (ericbottard) [inline spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: You're not up to date. Has been removed in https://github.com/spring-cloud/spring-cloud-dataflow/commit/efba70fdf5a49e1c126929770a16cb608bf97821

**Comment by ghillert** [comment]
Rebased.

File: spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml (hunk 6 above)
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: This is yet to be removed.

Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: will fix when merging.

**Comment by ilayaperumalg** [comment]
LGTM, merging.

**Comment by ilayaperumalg** [comment]
Squashed and merged as 35c6cfc494c40b577da5c0aa96bf89d20b8ea3f3 along with the cleanup of `tasks/deployments` endpoint references b4c7b6937533d743e772702395ec93463db0a170

**Pull request closed without merging**